mlflow/sagemaker/container/__init__.py

```diff
@@ -87,35 +99,52 @@ def _serve():
            "mlflow.sagemaker.container.scoring_server.wsgi:app").format(nworkers=cpu_count)
     bash_cmds.append(cmd)
     gunicorn = Popen(["/bin/bash", "-c", "; ".join(bash_cmds)])
-    signal.signal(signal.SIGTERM, lambda a, b: _sigterm_handler(nginx.pid, gunicorn.pid))
+    signal.signal(signal.SIGTERM, lambda a, b: _sigterm_handler(pids=[nginx.pid, gunicorn.pid]))
     # If either subprocess exits, so do we.
-    pids = set([nginx.pid, gunicorn.pid])
-    while True:
-        pid, _ = os.wait()
-        if pid in pids:
-            break
-    _sigterm_handler(nginx.pid, gunicorn.pid)
+    awaited_pids = _await_subprocess_exit_any(procs=[nginx, gunicorn])
+    _sigterm_handler(awaited_pids)
+
+
+def _serve_mleap(model):
+    mleap = Popen(["java", "-cp", "/opt/mlflow/mlflow/java/target/mlflow-java-"
+                   "{mlflow_version}-with-dependencies.jar".format(
+                        mlflow_version=mlflow.version.VERSION),
+                   "com.databricks.mlflow.sagemaker.ScoringServer",
+                   MODEL_PATH, DEFAULT_SAGEMAKER_SERVER_PORT])
+    signal.signal(signal.SIGTERM, lambda a, b: _sigterm_handler(pids=[mleap.pid]))
+    awaited_pids = _await_subprocess_exit_any(procs=[mleap])
+    _sigterm_handler(awaited_pids)
+
+
+def _container_includes_mlflow_source():
+    return os.path.isdir("/opt/mlflow")
 
 
 def _train():
     raise Exception("Train is not implemented.")
 
 
-def _sigterm_handler(nginx_pid, gunicorn_pid):
+def _await_subprocess_exit_any(procs):
+    pids = [proc.pid for proc in procs]
+    while True:
+        pid, _ = os.wait()
+        if pid in pids:
+            break
+    return pids
+
+
+def _sigterm_handler(pids):
     """
     Cleanup when terminating.
 
     Attempt to kill all launched processes and exit.
 
     """
     print("Got sigterm signal, exiting.")
-    try:
-        os.kill(nginx_pid, signal.SIGQUIT)
-    except OSError:
-        pass
-    try:
-        os.kill(gunicorn_pid, signal.SIGTERM)
```

Comment:
Why SIGQUIT v SIGTERM? Now they are both SIGQUIT -- which is the core dump signal. I would kind of expect SIGTERM to be sent instead?

Reply:
I copied the code from example. I guess the author wanted to have  core dump if something happened? I think keeping only sigterm makes sense.

Reply:
Awesome, sigterm sounds good.